extensions/guacamole-auth-ldap/src/main/java/org/apache/guacamole/auth/ldap/AuthenticationProviderService.java

```diff
@@ -307,16 +312,39 @@ public LDAPAuthenticatedUser authenticateUser(Credentials credentials)
 
     }
 
+    /**
+     * Returns parameter current ldap domain token generated from user credentials
+     * @param credentials
+     *     The credentials used for authentication.
+     *
+     * @return
+     *     Domain name by splitting login username or null if no domain is detected.
+     */
+    private String getDomainToken(Credentials credentials) {
+        String ldapDomainName = null;
+        if (credentials.getUsername().contains("\\")) {
+            ldapDomainName = credentials.getUsername().split("\\\\")[0];
+        }
+        else if (credentials.getUsername().contains("@")) {
+            ldapDomainName = credentials.getUsername().split("@")[1];
+        }
+        return ldapDomainName;
+    }
```

Comment:
Hearkening back to the initial review, I'm still on the fence about this for the same reason:

> I don't think ... that we should make specific assumptions about the format here when the LDAP handling of username format is so flexible.

That said, I do see the utility of having this sort of token just be available, given that this sort of matching is going to be so common.

If we're going to have a `${LDAP_DOMAIN}` token that is always defined for any successful login that follows the `DOMAIN\username` or `username@domain` formats, I think we should do this parsing a bit more efficiently. Currently, the algorithm here is:

1. Search the username string for the first occurrence of a `\` character.
2. If such a character is found, search the username string again, this time splitting the string into an array. Discard everything but the first element in that array.
3. Search the username string yet again, this time for the first occurrence of a `@` character.
4. If such a character is found, search the username string _again_, this time splitting the string into an array. Discard everything but the first element in that array.

This is suboptimal, as:

* It potentially searches the username string 4 times.
* It unnecessarily splits the string into an array, rather than just parsing the interesting part  of the string directly.

Instead, I'd recommend defining and using a `Pattern` that matches _either_ of these formats, extracting the domain. This would reduce things down to a single operation that performs only one search and directly extracts the domain at the same time.

Reply:
Agree, I will do that.